### PR TITLE
Fixed #29103 -- Fixed SchemaEditor.quote_value() for unicode strings on MySQL.

### PR DESCRIPTION
**Pull request by claudep**
https://code.djangoproject.com/ticket/29103

### CODE REVIEW

File: tests/backends/mysql/tests.py [hunk 1]
```diff
@@ -89,3 +89,19 @@ def test_isolation_level_validation(self):
         )
         with self.assertRaisesMessage(ImproperlyConfigured, msg):
             new_connection.cursor()
+
+
+@unittest.skipUnless(connection.vendor == 'mysql', 'MySQL tests')
```
Comment (timgraham) [inline tests/backends/mysql/tests.py]: I'd put this in a separate `test_schema.py` file -- but is it really MySQL-specific? I guess some database might quote those values differently but I didn't check. I would put the non-regression tests (everything except the bytestring?) in a separate commit that precedes the bug fix so it's more clear exactly which case this fixes. `return self.connection.connection.escape(value, self.connection.connection.encoders)` might be considered as a separate refactoring commit also.

File: tests/backends/mysql/tests.py [hunk 2]
```diff
@@ -89,3 +89,19 @@ def test_isolation_level_validation(self):
...
+            ('', "''"),
+            (b'', b"''"),
+            ("l'hôte", r"'l\'hôte'"),
+            ('an "id"', r"'an \"id\"'"),
+        )
+        with connection.schema_editor() as editor:
+            for value, expected in possible_params:
+                with self.subTest(value=value):
+                    quoted = editor.quote_value(value)
+                    self.assertEqual(quoted, expected)
```
Comment (timgraham) [inline tests/backends/mysql/tests.py]: `self.assertEqual(editor.quote_value(value), expected)` fits on one line, I think.

File: tests/backends/mysql/tests.py [hunk 3]
```diff
@@ -89,3 +89,19 @@ def test_isolation_level_validation(self):
         )
         with self.assertRaisesMessage(ImproperlyConfigured, msg):
             new_connection.cursor()
+
+
+@unittest.skipUnless(connection.vendor == 'mysql', 'MySQL tests')
+class SchemaEditorTests(TestCase):
+    def test_quote_value(self):
+        possible_params = (
```
Comment (timgraham) [inline tests/backends/mysql/tests.py]: possible_params -> tests 
is consistent with most other subTests

**Comment by claudep** [comment]
I found another test targeting the same sort of output in `SchemaLoggerTests`, what do you think?
To be tested on Oracle.

**Comment by timgraham** [comment]
That doesn't look like the appropriate location for the test consider that test's original purpose (1cb65b8a77bd1b15f094b5857468c107784aad90).

**Comment by claudep** [comment]
This still needs work, as even if I prevented an existing crash with my code, I'm still unsure about the exact binary syntax we should output in the collected SQL.

**Comment by carltongibson** [comment]
Hi @claudep: 

> This still needs work...

Can I ask you to elaborate please? I don't follow you. (If it weren't for your comment I'd say this is good to go...) 

The CI failure's here don't reproduce locally (and they've expired so I can't see the cause).

**Comment by claudep** [comment]
I just force pushed to rerun the CI. The tricky issue is about binary strings. But as binary content in queries is really a corner case, I would suggest not to block the ticket, fix the issue by removing the binary from the test, and maybe add another ticket for the binary use case.

**Comment by carltongibson** [comment]
> ... fix the issue by removing the binary from the test...

But the test passes. (That's what I'm not quite seeing. Sorry for being slow.)

**Comment by carltongibson** [comment]
Ah, but they fail on Jenkins (where not locally)... OK Sorry, let me look into that.

**Comment by claudep** [comment]
Ah, maybe my initial test was testing a bit more precisely the SQL output. The current test result for MySQL is `'SELECT * FROM foo WHERE id in (42, \'café\', b"\'\\xe4\'");'` which is probably not an instruction that MySQL will accept on the command line. The fact that it doesn't crash is already a progress in my sight.

**Comment by carltongibson** [comment]
The failures are totally unrelated: 

```
AssertionError: Database queries aren't allowed in SimpleTestCase. Either use TestCase or TransactionTestCase to ensure proper test isolation or set SchemaLoggerTests.allow_database_queries to True to silence this failure.
```

**Comment by carltongibson** [comment]
My `HEAD` (from pulling the PR) was showing a different commit hash to the one shown here (but the same patch diff) so I've rebased on `master` and pushed again.

If that doesn't solve the build failure I'm going to call in the cavalry. 🙂

File: tests/logging_tests/tests.py [hunk 4]
```diff
@@ -463,7 +463,7 @@ def test_custom_logging(self):
         self.assertOutput(out, "System check identified no issues (0 silenced).")
 
 
-class SchemaLoggerTests(SimpleTestCase):
+class SchemaLoggerTests(TestCase):
```
Comment (carltongibson) [inline tests/logging_tests/tests.py]: `quote_value` now requires a connection, so we need this change too.

**Comment by carltongibson** [comment]
@claudep I adjusted the failing test (as per the instructions on the trace 🙂). Are you happy with the change? If so I'll mark it RFC. 

Thanks for the effort!

**Comment by claudep** [comment]
Thanks for the fix!

**Comment by timgraham** [comment]
So, we will open a separate ticket for the issue that `b"'\xe4'"` is output in the SQL string? (if it can be fixed) Maybe it makes sense to add a test for `quote_value()`, even if it has to be MySQL-specific. I'm not convinced the proposed test does a good job at demonstrating that the original issue in the ticket is fixed (and preventing a regression there).

**Comment by claudep** [comment]
Added test for `quote_value` in backend tests.

**Pull request merged**